packages/foundry/contracts/FYToken.sol

```diff
@@ -118,15 +140,54 @@ contract FYToken is IFYToken, IERC3156FlashLender, AccessControl, ERC20Permit, C
         accrual_ = accrual_ >= 1e18 ? accrual_ : 1e18; // The accrual can't be below 1 (with 18 decimals)
     }
 
+    ///@dev returns the maximum redeemable amount for the address holder in terms of the principal
+    function maxRedeem(address holder) public view returns (uint256 maxPrincipalAmount) {
+        return _balanceOf[holder];
+    }
+
+    ///@dev returns the amount of underlying redeemable in terms of the principal
+    function previewRedeem(uint256 principalAmount) public view returns (uint256 underlyingAmount) {
+        return _previewRedeem(principalAmount);
+    }
+
+    ///@dev returns the amount of underlying redeemable in terms of the principal
+    function _previewRedeem(uint256 principalAmount) private view returns (uint256 underlyingAmount) {
+        return _convertToUnderlying(principalAmount);
+    }
+
     /// @dev Burn fyToken after maturity for an amount that increases according to `chi`
     /// If `amount` is 0, the contract will redeem instead the fyToken balance of this contract. Useful for batches.
-    function redeem(address to, uint256 amount) external override afterMaturity returns (uint256 redeemed) {
-        uint256 amount_ = (amount == 0) ? _balanceOf[address(this)] : amount;
-        _burn(msg.sender, amount_);
-        redeemed = amount_.wmul(_accrual());
-        join.exit(to, redeemed.u128());
+    function redeem(uint256 principalAmount, address receiver, address holder) external override afterMaturity returns (uint256 underlyingAmount) {
+        uint256 amount_ = (principalAmount == 0) ? _balanceOf[address(this)] : principalAmount;
```

Comment:
upon further thought, i kinda of like the overload, because if you use type(uint).max (or zero) for something else, then it kinda breaks the whole concept that preview should show exactly what is being redeemed (but at the same time, preview should ignore how much the caller actually holds)

Reply:
it sb a overloaded external fn that looks up the callers bal and then passes it in to the users ...

if we go this route, then redeem / withdraw sb external fns that call internal fns -- this is different than the implementation we did but i think thats right https://github.com/yieldprotocol/ERC5095/pull/1/files

Reply:
So, this is one of my dirty hacks, to allow the fyToken to redeem what was sent to it, without necessarily having to know how much is it.

In a batch, when we burn some pool tokens, we might want to redeem the fyToken part if it is mature, but it will be impossible to know exactly how much is that.

I think that a second `function redeem(address receiver) external afterMaturity returns (uint256 underlyingAmount)` is indeed much cleaner. The Ladle doesn't ever call `redeem` itself (please check!), so we could get away with refactoring it and letting the frontend guys know.